### PR TITLE
expression: fix error PushDownNot for contiguous NOT

### LINKED ISSUE #15725: Double negation causes incorrect result
Consider the following statements:

```sql
CREATE TABLE t0(c0 INT);
INSERT INTO t0(c0) VALUES (2);
SELECT t0.c0 FROM t0 WHERE (NOT NOT t0.c0) = t0.c0; -- expected: {}, actual: {2}
```
Unexpectedly, a row is fetched. The predicate `(NOT NOT t0.c0) = t0.c0`. should evaluate to false. I suspect that the double negation is removed, which is incorrect, since `NOT NOT t0.c0` should evaluate to `1`, and not to `2`.

Here is the environment that I used:
```sql
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                           |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.12
Git Commit Hash: 8c4696b3f3408c61dd7454204ddd67c93501609a
Git Branch: heads/refs/tags/v3.0.12
UTC Build Time: 2020-03-16 09:56:22
GoVersion: go version go1.13 linux/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #15725  

Problem Summary:
The issue exposes 2 problems:
1. double negation is handled wrongly in PushDownNot. This problem only appears in release-3.0 and release-3.1 because of `problem 2`.

2. If `not` appears even number of times in an expression tree, the `not` can not be eliminated successfully. `(not not not not a)` can be optimized to `(not not a)` here. 
``` sql
tidb> desc select * from t where (not not not not a)=a;
+-------------------------+---------+-----------+---------------+--------------------------------------------+
| id                      | estRows | task      | access object | operator info                              |
+-------------------------+---------+-----------+---------------+--------------------------------------------+
| TableReader_7           | 0.80    | root      |               | data:Selection_6                           |
| └─Selection_6           | 0.80    | cop[tikv] |               | eq(not(not(not(not(test.t.a)))), test.t.a) |
|   └─TableFullScan_5     | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo             |
+-------------------------+---------+-----------+---------------+--------------------------------------------+
3 rows in set (0.00 sec)
```

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Do not eliminate the innermost `NOT`.

How it Works:

### Related changes

N/A
I'll push specified PR to release branch.

### Check List 

Tests 

- Unit test
- Integration test

Side effects

N/A

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
#16094 is the same bug-fix on release-3.0

File: expression/util.go [hunk 1]
```diff
@@ -369,12 +387,30 @@ func pushNotAcrossArgs(ctx sessionctx.Context, exprs []Expression, not bool) ([]
...
 	if f, ok := expr.(*ScalarFunction); ok {
 		switch f.FuncName.L {
 		case ast.UnaryNot:
-			return pushNotAcrossExpr(f.GetCtx(), f.GetArgs()[0], !not)
+			var childExpr Expression
+			// UnaryNot only returns 0/1/NULL, thus we should not eliminate the
+			// innermost NOT if the arg is not a logical operator.
+			switch child := f.GetArgs()[0].(type) {
+			case *ScalarFunction:
+				if _, isLogicalOp := logicalOps[child.FuncName.L]; !isLogicalOp {
```
Comment (eurekaka) [inline expression/util.go:403]: Can we wrap the argument of innermost Not with a `IsTruth` function, so we can remove all even numbers of Not?

File: expression/expression.go [hunk 2]
```diff
@@ -1111,7 +1111,11 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
 // TODO: remove this function. ScalarFunction should be newed in one place.
 func wrapWithIsTrue(ctx sessionctx.Context, keepNull bool, arg Expression) (Expression, error) {
 	if arg.GetType().EvalType() == types.ETInt {
-		return arg, nil
+		if child, ok := arg.(*ScalarFunction); ok {
```
Comment (eurekaka) [inline expression/expression.go:1114]: Shall we add a context flag to indicate that we are in `PushDownNot` and then do this check? then we can avoid incurring unnecessary `IsTrue()` functions like those in the above explain test result.

File: expression/util.go [hunk 3]
```diff
@@ -369,12 +387,24 @@ func pushNotAcrossArgs(ctx sessionctx.Context, exprs []Expression, not bool) ([]
...
+func pushNotAcrossExpr(ctx sessionctx.Context, expr Expression, not bool) (_ Expression, changed bool) {
 	if f, ok := expr.(*ScalarFunction); ok {
 		switch f.FuncName.L {
 		case ast.UnaryNot:
-			return pushNotAcrossExpr(f.GetCtx(), f.GetArgs()[0], !not)
+			child, err := wrapWithIsTrue(ctx, true, f.GetArgs()[0])
+			if err != nil {
+				return expr, false
+			}
+			var childExpr Expression
```
[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply expression/util.go:402]: No, if so, `not` and `changed` both need to be checked.
This is an example if we replace it.
``` go
tidb> desc select a from t where (not not not not (a+1)) =a ;
+-------------------------+---------+-----------+---------------+-----------------------------------------------------+
| id                      | estRows | task      | access object | operator info                                       |
+-------------------------+---------+-----------+---------------+-----------------------------------------------------+
| TableReader_7           | 2.40    | root      |               | data:Selection_6                                    |
| └─Selection_6           | 2.40    | cop[tikv] |               | eq(not(not(not(not(plus(test.t.a, 1))))), test.t.a) |
|   └─TableFullScan_5     | 3.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo                      |
+-------------------------+---------+-----------+---------------+-----------------------------------------------------+
3 rows in set (0.00 sec)
```

**Comment by XuHuaiyu** [comment]
unit-test fail because of https://github.com/pingcap/tidb/issues/15979

**Pull request merged**